src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -322,27 +308,40 @@ private static bool ValidateName(string name, bool targetingAttributes, ErrorSin
 
             var validName = TryValidateName(
                 name,
-                whitespaceError: Resources.FormatHtmlTargetElementAttribute_NameCannotBeNullOrWhitespace(targetName),
-                characterErrorBuilder: (invalidCharacter) =>
-                    Resources.FormatHtmlTargetElementAttribute_InvalidName(
-                        targetName.ToLower(),
-                        name,
-                        invalidCharacter),
+                targetName,
                 errorSink: errorSink);
 
             return validName;
         }
 
         private static bool TryValidateName(
             string name,
-            string whitespaceError,
-            Func<char, string> characterErrorBuilder,
+            string tagHelperName,
+            ErrorSink errorSink)
+        {
+            return TryValidateName(
+                name,
+                tagHelperName,
+                false,
+                errorSink);
+            
+        }
+
+        private static bool TryValidateName(
```

Comment:
It's time to split this in two or more e.g. `TryValidateAttributeName()` and `TryValidateRestrictChildrenAttributeName()`. Each method can then use the appropriate resources when errors occur. Not that much of the code is truly separate from the error formatting.

Alternatively, separate the whitespace check (two methods or inline at the 3 call sites) from an invalid character check e.g. `private int FindInvalidCharacter(string name)` which returns the index of an invalid `char` or `-1` when all is good. This alternative nicely separates looking for low-level problems from adding errors about those problems.